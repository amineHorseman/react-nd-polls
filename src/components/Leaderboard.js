import { selectAuthedUserId } from "../features/auth/authSlice";
import { selectAllUsers } from "../features/users/usersSlice";
import { useUsersRanking } from "../hooks/useUsersRanking";
import { useSelector } from "react-redux";

const Leaderboard = () => {
    const users = useSelector(selectAllUsers);
    const authedUserId = useSelector(selectAuthedUserId);
    const { scores, authedUserRank } = useUsersRanking(authedUserId);

    return (
        <div className="container mt-4 col-lg-10">
            <h2 className="mb-4">Leaderboard</h2>
            <div className="table-responsive">
                <table className="table table-hover">
                    <thead className="table-dark">
                        <tr>
                            <th>Rank</th>
                            <th>User</th>
                            <th>Total Score</th>
                            <th>Questions</th>
                            <th>Answers</th>
                        </tr>
                    </thead>
                    <tbody>
                        { scores.map(({id, score}, index) => (
                            <tr key={id}
                                className={index === authedUserRank ? 'table-light' : ''}>
                                <td>{index + 1}</td>
                                <td>{users[id].name} <span className="text-muted">(@{id})</span></td>
                                <td><strong>{score}</strong></td>
                                <td>{users[id].questions.length}</td>
                                <td>{Object.keys(users[id].answers).length}</td>
                            </tr>
                        ))}
                    </tbody>
                </table>
            </div>
        </div>
    );
};

export default Leaderboard;
