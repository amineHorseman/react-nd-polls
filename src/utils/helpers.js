import { FaCircleUser } from "react-icons/fa6"

export function displayAvatar(avatarURL) {
    // Display user avatar, or a profile icon if URL is null
    return (
        <span className="avatar">
        { avatarURL === null ? 
            <FaCircleUser /> : 
            <img src={avatarURL} className="rounded-circle" alt="User avatar"/>
        }
        </span>
    );
}

export function displayElapsedTime(timestamp) {
    // Display elapsed time since 'timestamp' in a readable text format
    const elapsedTimeString = getElapsedTime(timestamp);
    return <span className="text-muted">{ elapsedTimeString }</span>;
}

export function getElapsedTime(timestamp) {
    // Compute the elapsed time since 'timestamp', and returns it as a readable string
    const elapsedTimestamp = Date.now() - timestamp; 
    const minutes = Math.floor(elapsedTimestamp / (1000 * 60));
    const hours = Math.floor(minutes / 60);
    const days = Math.floor(hours / 24);
    const months = Math.floor(days / 30);
    const years = Math.floor(months / 12);

    if (years > 1)
        return `${years} year(s) ago`
    else if (months > 1)
            return `${months} days ago`
    else if (days > 1)
        return `${days} days ago`
    else if (hours > 1)
        return `${hours} hours ago`
    else if (minutes > 1)
        return `${minutes} minutes ago`
    else
        return 'less than a minute ago'
}

export const handleFormChange = (e, field, setFormState) => {
    setFormState(formState => ({...formState, 
        [field]: e.target.value,
        error: ''}));
}
