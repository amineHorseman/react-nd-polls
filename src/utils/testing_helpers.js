import { fetchQuestions } from "../features/questions/questionsSlice";
import { fetchUsers } from "../features/users/usersSlice";
import { loginUser } from "../features/auth/authSlice";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import { Provider } from "react-redux";
import store from '../app/store';

export function wrapAndRenderComponent(component) {
    return render(
        <BrowserRouter>
            <Provider store={store}>
                {component}
            </Provider>
        </BrowserRouter>
    );
}

export async function loadInitialDataForTesting() {
    await store.dispatch(fetchUsers());
    await store.dispatch(fetchQuestions());
}

export async function loginUserForTesting(username) {
    await store.dispatch(loginUser(username));
}
