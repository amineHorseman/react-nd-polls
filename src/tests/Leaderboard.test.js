import '@testing-library/jest-dom';
import { loadInitialDataForTesting } from "../utils/testing_helpers";
import { wrapAndRenderComponent } from '../utils/testing_helpers';
import { loginUserForTesting } from "../utils/testing_helpers";
import Leaderboard from "../components/Leaderboard";

describe("Test for Leaderboard component", () => {

    test("Component matches snapshot with user mtsamis", async () => {
        await loadInitialDataForTesting();
        await loginUserForTesting('mtsamis');
        const view = wrapAndRenderComponent(<Leaderboard />);
        expect(view).toMatchSnapshot();
    });
});
