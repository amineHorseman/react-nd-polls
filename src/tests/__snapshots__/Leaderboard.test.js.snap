// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Leaderboard component Component matches snapshot with user mtsamis 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container mt-4 col-lg-10"
      >
        <h2
          class="mb-4"
        >
          Leaderboard
        </h2>
        <div
          class="table-responsive"
        >
          <table
            class="table table-hover"
          >
            <thead
              class="table-dark"
            >
              <tr>
                <th>
                  Rank
                </th>
                <th>
                  User
                </th>
                <th>
                  Total Score
                </th>
                <th>
                  Questions
                </th>
                <th>
                  Answers
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class=""
              >
                <td>
                  1
                </td>
                <td>
                  Sarah Edo
                   
                  <span
                    class="text-muted"
                  >
                    (@
                    sarahedo
                    )
                  </span>
                </td>
                <td>
                  <strong>
                    6
                  </strong>
                </td>
                <td>
                  2
                </td>
                <td>
                  4
                </td>
              </tr>
              <tr
                class="table-light"
              >
                <td>
                  2
                </td>
                <td>
                  Mike Tsamis
                   
                  <span
                    class="text-muted"
                  >
                    (@
                    mtsamis
                    )
                  </span>
                </td>
                <td>
                  <strong>
                    5
                  </strong>
                </td>
                <td>
                  2
                </td>
                <td>
                  3
                </td>
              </tr>
              <tr
                class=""
              >
                <td>
                  3
                </td>
                <td>
                  Tyler McGinnis
                   
                  <span
                    class="text-muted"
                  >
                    (@
                    tylermcginnis
                    )
                  </span>
                </td>
                <td>
                  <strong>
                    4
                  </strong>
                </td>
                <td>
                  2
                </td>
                <td>
                  2
                </td>
              </tr>
              <tr
                class=""
              >
                <td>
                  4
                </td>
                <td>
                  Zenobia Oshikanlu
                   
                  <span
                    class="text-muted"
                  >
                    (@
                    zoshikanlu
                    )
                  </span>
                </td>
                <td>
                  <strong>
                    1
                  </strong>
                </td>
                <td>
                  0
                </td>
                <td>
                  1
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container mt-4 col-lg-10"
    >
      <h2
        class="mb-4"
      >
        Leaderboard
      </h2>
      <div
        class="table-responsive"
      >
        <table
          class="table table-hover"
        >
          <thead
            class="table-dark"
          >
            <tr>
              <th>
                Rank
              </th>
              <th>
                User
              </th>
              <th>
                Total Score
              </th>
              <th>
                Questions
              </th>
              <th>
                Answers
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class=""
            >
              <td>
                1
              </td>
              <td>
                Sarah Edo
                 
                <span
                  class="text-muted"
                >
                  (@
                  sarahedo
                  )
                </span>
              </td>
              <td>
                <strong>
                  6
                </strong>
              </td>
              <td>
                2
              </td>
              <td>
                4
              </td>
            </tr>
            <tr
              class="table-light"
            >
              <td>
                2
              </td>
              <td>
                Mike Tsamis
                 
                <span
                  class="text-muted"
                >
                  (@
                  mtsamis
                  )
                </span>
              </td>
              <td>
                <strong>
                  5
                </strong>
              </td>
              <td>
                2
              </td>
              <td>
                3
              </td>
            </tr>
            <tr
              class=""
            >
              <td>
                3
              </td>
              <td>
                Tyler McGinnis
                 
                <span
                  class="text-muted"
                >
                  (@
                  tylermcginnis
                  )
                </span>
              </td>
              <td>
                <strong>
                  4
                </strong>
              </td>
              <td>
                2
              </td>
              <td>
                2
              </td>
            </tr>
            <tr
              class=""
            >
              <td>
                4
              </td>
              <td>
                Zenobia Oshikanlu
                 
                <span
                  class="text-muted"
                >
                  (@
                  zoshikanlu
                  )
                </span>
              </td>
              <td>
                <strong>
                  1
                </strong>
              </td>
              <td>
                0
              </td>
              <td>
                1
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
